src/lib/containers/entity_finder/tree/VirtualizedTree.tsx

```diff
@@ -0,0 +1,510 @@
+import { Skeleton } from '@material-ui/lab'
+import { Map } from 'immutable'
+import { cloneDeep } from 'lodash'
+import React, { FC, useCallback, useEffect, useState } from 'react'
+import { useInView } from 'react-intersection-observer'
+import AutoSizer from 'react-virtualized-auto-sizer'
+import {
+  TreeWalker,
+  TreeWalkerValue,
+  VariableSizeNodeData,
+  VariableSizeNodePublicState,
+  VariableSizeTree,
+} from 'react-vtree'
+import { NodeComponentProps } from 'react-vtree/dist/es/Tree'
+import { SynapseClient } from '../../../utils'
+import {
+  getEntityTypeFromHeader,
+  isContainerType,
+} from '../../../utils/functions/EntityTypeUtils'
+import { useSynapseContext } from '../../../utils/SynapseContext'
+import {
+  EntityHeader,
+  EntityType,
+  ProjectHeader,
+} from '../../../utils/synapseTypes'
+import { Writable } from '../../../utils/types/Writable'
+import { EntityBadgeIcons } from '../../EntityBadgeIcons'
+import { EntityTypeIcon } from '../../EntityIcon'
+import { SynapseSpinner } from '../../LoadingScreen'
+
+export enum EntityTreeNodeType {
+  /** The tree component's appearance and interactions will facilitate selection. Nodes will be larger and styles will indicate primary selection */
+  SELECT,
+  /** The tree component's appearance and interactions will facilitate browsing. Nodes will be smaller and styles will indicate secondary selection */
+  BROWSE,
+}
+
+type EntityFinderHeader =
+  | Pick<EntityHeader, 'id' | 'name' | 'type'>
+  | ProjectHeader
+
+type NodeChildren = Readonly<{
+  /** The node's children. If undefined, then children have not been fetched. */
+  children?: EntityHeaderNode[]
+  /** The token used to fetch the next page of children for this entity. If this property is nullish and children is defined, then all children have been fetched. */
+  childrenNextPageToken?: string | null
+}>
+
+type EntityHeaderNode = EntityFinderHeader & NodeChildren
+type PaginationNode = { __paginationNode: true }
+
+type TreeNode = EntityHeaderNode | RootNodeConfiguration | PaginationNode
+
+function getTreeNodeType(
+  node: TreeNode,
+): 'entityHeader' | 'rootNodeConfiguration' | 'pagination' {
+  if ('__paginationNode' in node) {
+    return 'pagination'
+  }
+  return 'id' in node ? 'entityHeader' : 'rootNodeConfiguration'
+}
+
+function isPaginationNode(node: TreeNode): node is PaginationNode {
+  return getTreeNodeType(node) === 'pagination'
+}
+
+function isEntityHeaderNode(node: TreeNode): node is EntityHeaderNode {
+  return getTreeNodeType(node) === 'entityHeader'
+}
+
+function isRootNodeConfiguration(
+  node: TreeNode,
+): node is RootNodeConfiguration {
+  return getTreeNodeType(node) === 'rootNodeConfiguration'
+}
+
+function hasMoreChildren(node: TreeNode) {
+  if (isPaginationNode(node)) {
+    return false
```

Comment:
TODO: This is at odds with the data.hasMoreChildren field. This is always true for pagination nodes. Maybe remove the field and always use this function, renamed slightly?